### ISSUE TITLE
[SPIR-V] Loop unrolling fails with invalid block merging

### ISSUE THREAD

**Issue by LukasBanana** [microsoft/DirectXShaderCompiler#2985]
Hi, I boiled down a shader to a minimum example that causes the latest revision of DXC to fail during SPIR-V loop unrolling:

```hlsl
Texture2D<float> InputTexture;
RWTexture2D<float> OutputTexture;

float Lookup(Texture2D<float> Tex) {
	return Tex.Load(0);
}

float Conditional(float d) {
	if (d < 1) {
		return 1;
	} else {
		return 0;
	}
}

[numthreads(1, 1, 1)]
void MainCS() {
	float InputValue = Lookup(InputTexture);
	float OutValue = 0;
	[unroll]
	for (uint i = 0; i < 2; i++) {
		float Offset = abs(i);
		float Value = Conditional(Offset);
		OutValue += (InputValue * Value);
	}
	OutputTexture[(uint2)0] = OutValue;
}
```
The bug disappears if either the intrinsic call `abs` is removed, or the function call `Lookup` is manually inlined. Otherwise, the following error is reported:
```
fatal error: generated SPIR-V is invalid: Block 34[%34] is already a merge block for another header
  %MainCS = OpFunction %void None %20
```
My command line for `dxcompiler` is:
```
-E MainCS -T cs_6_0 -spirv -Fo LoopUnrollBug.spv LoopUnrollBug.hlsl
```
The bug also disappears when the loop unrolling optimization pass is removed from `Optimizer::RegisterPerformancePasses()` in SPIRV-Tools:
```cpp
.RegisterPass(CreateLoopUnrollPass(true))
```
Thanks and kind regards
Lukas

**Comment by ehsannas** [comment]
This has just been fixed in SPIRV-Tools today (https://github.com/KhronosGroup/SPIRV-Tools/issues/3441). Will pick up the fix in DXC tomorrow.

**Comment by LukasBanana** [comment]
Perfect, thanks for the quick reply.
